From 3ef1c653ba92642736ff4fc4e2eb0395e12e7e81 Mon Sep 17 00:00:00 2001
From: Limeng <Meng.Li@windriver.com>
Date: Tue, 19 May 2020 18:59:32 +0800
Subject: [PATCH] driver: gpu: dcss: disable and clear DTG module interrupt
 during initialization

When run kexec/kdump feature, the second kernel hangs during
loading DCSS(Display controller subsystem) driver. Because there
are some DTG(Display Timing Generator) interrupts that are still
in pending when the first kernel quits.
Therefore, fix this issue by adding interrupt disable and clear
operation to make sure there is no unsteable interrupt.

Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dcss/dcss-dtg.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/gpu/imx/dcss/dcss-dtg.c b/drivers/gpu/imx/dcss/dcss-dtg.c
index f90fe86ba35c..227538c31fc5 100644
--- a/drivers/gpu/imx/dcss/dcss-dtg.c
+++ b/drivers/gpu/imx/dcss/dcss-dtg.c
@@ -260,6 +260,18 @@ int dcss_dtg_init(struct dcss_soc *dcss, unsigned long dtg_base)
 	dtg->control_status |= OVL_DATA_MODE | BLENDER_VIDEO_ALPHA_SEL |
 		((dtg->alpha << DEFAULT_FG_ALPHA_POS) & DEFAULT_FG_ALPHA_MASK);
 
+	/* Disable interrupt */
+	dcss_update(0, LINE0_IRQ, dtg->base_reg + DCSS_DTG_INT_MASK);
+	dcss_update(0, LINE1_IRQ, dtg->base_reg + DCSS_DTG_INT_MASK);
+	dcss_update(0, LINE2_IRQ, dtg->base_reg + DCSS_DTG_INT_MASK);
+	dcss_update(0, LINE3_IRQ, dtg->base_reg + DCSS_DTG_INT_MASK);
+
+	/* Clear interrupt */
+	dcss_update(LINE0_IRQ, LINE0_IRQ, dtg->base_reg + DCSS_DTG_INT_CONTROL);
+	dcss_update(LINE1_IRQ, LINE1_IRQ, dtg->base_reg + DCSS_DTG_INT_CONTROL);
+	dcss_update(LINE2_IRQ, LINE2_IRQ, dtg->base_reg + DCSS_DTG_INT_CONTROL);
+	dcss_update(LINE3_IRQ, LINE3_IRQ, dtg->base_reg + DCSS_DTG_INT_CONTROL);
+
 	return dcss_dtg_irq_config(dtg);
 }
 
-- 
2.17.1

